Proposed API change

michael-petersen opened the pull request:
This version of `lintsampler` has a streamlined interface that more closely matches Pythonic patterns. At the simplest, we can now do something like draw 6 samples from a box:

  ```python
  >>> x = np.linspace(10,20,2)
  >>> y = np.linspace(100,200,2)
  >>> z = np.linspace(1000,2000,2)
  >>> def rndmpdf(X): return np.random.uniform(size=X.shape[0])
  >>> LS = LintSampler(rndmpdf,cells=(x,y,z)).sample(6)
  array([[  12.63103673,  186.7514952 , 1716.6187807 ],
         [  14.67375968,  116.20984414, 1557.59629547],
         [  11.47055697,  178.41650558, 1592.18260186],
         [  12.41780309,  105.28009531, 1436.39525998],
         [  13.44764381,  152.57623376, 1880.55963378],
         [  18.5522151 ,  133.87092063, 1558.85620176]])
  ```

which entails constructing a probability density function (here built as a vectorised input to `numpy.random.uniform`), and defining the cell, _before_ calling for a number of samples (here, 6).

The wrapper class makes both the special case of a grid and the more flexible case of a set of cells fall under one call structure.

michael-petersen committed:
first wrapper implementation

michael-petersen committed:
implement 1d gridsample case; promote pdf to init

michael-petersen committed:
add cell handling input

michael-petersen committed:
change grid to cells

michael-petersen committed:
change call structure; add docstrings

michael-petersen committed:
add grid input functionality

michael-petersen committed:
doc improvement (usage.md)

michael-petersen commented:
It would also be nice to implement the flexibility option for the pdf input, such that the pdf function doesn't need to be vectorised.

michael-petersen committed:
improve rnginput flexibility; allow for non-vectorised pdfs

michael-petersen commented:
The new support for non-vectorised pdfs (which is now the default) means that a minimal example to draw from a 3d prism looks like:
```
import numpy as np
from lintsampler import LintSampler
 
x = np.linspace(10,20,2)
y = np.linspace(100,200,2)
z = np.linspace(1000,2000,2)
 
# default: a pdf that takes the same number of arguments as dimensions
def rndmpdf_nonvec(x,y,z):
    return np.random.uniform()
 
LintSampler(rndmpdf_nonvec,cells=(x,y,z)).sample()
```
But the call also works to use a vectorised pdf, for performance, if the `vectorizedpdf` flag is set:
```
def rndmpdf(X):
    return np.random.uniform(size=X.shape[0])

LintSampler(rndmpdf,cells=(x,y,z),vectorizedpdf=True).sample()
```

michael-petersen committed:
refactor pdf calls to their own functions

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+        ----------
+        self : LintSampler
+            The LintSampler instance.
+        cells : tuple of arrays or list of tuples of arrays
```
Ditto comment above: `cells` can also be lone array

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+
+        # check the validity of the input
+        if len(cells)==0:
+            raise ValueError("LintSampler._setgrid: you must specify an evaluation domain with a tuple of arrays or a list of tuples of arrays. See documentation for details.")
```
```suggestion
            raise ValueError("LintSampler._setgrid: you must specify an evaluation domain with a single array, a tuple of arrays or a list of tuples of arrays. See documentation for details.")
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+    samples within the grid.
+    """
+
+    def __init__(self, pdf, cells=(), rnginput=None, vectorizedpdf=False):
```
Here a default value (empty tuple) is given to cells, but (I believe) this default value would throw an error, should we remove the default value?
```suggestion
    def __init__(self, pdf, cells, rnginput=None, vectorizedpdf=False):
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+        cells : tuple of arrays or list of tuples of arrays
+            If a tuple of arrays, the boundary values for a grid with dimensionality of the length
+            of the tuple. If a list of tuples of arrays, the boundary values for an arbitrary
+            number of grids with dimensionality the length of the tuples.
```
I think `cells` can also be just an array? (as in the first example in the docstring)
```suggestion
        cells : single array, tuple of arrays or list of tuples of arrays
            If a single array, the boundary values for a 1D grid. If a tuple of arrays, the
            boundary values for a grid with dimensionality of the length of the tuple. If
            a list of tuples of arrays, the boundary values for an arbitrary number of
            grids with dimensionality the length of the tuples.
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+
+        # 0. are we in the 1d case? the input is just a single tuple or 1d array.
+        # i.e. cells = np.linspace(-12,12,100)
+        if (isinstance(cells[0],float) | (isinstance(cells[0],int))): # do we need the integer option here? I think no?
```
Here, maybe we want to catch all cases of 1D iterables, e.g. 1D array, tuple of numbers, list of numbers?
```suggestion
        if hasattr(cells, "__len__") and not hasattr(cells[0], "__len__")
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+
+            # set the arrays
+            # we are being passed the array directly
+            self.edgearrays = cells
```
I'm not sure if this is important for later, but in case the user provides a 1D tuple/list, maybe here we cast it to an array?
```suggestion
            self.edgearrays = np.array(cells)
```

michael-petersen replied:
Yes, probably safest.

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+
+        # 1. tuples defining the array -> make arrays, pass to gridsample
+        # i.e. cells = (np.linspace(-12,12,100),np.linspace(-4,4,50))
+        if isinstance(cells,tuple):
```
Should we make this an `elif` so that it doesn't catch and override the case of a 1D tuple?
```suggestion
        elif isinstance(cells,tuple):
```

michael-petersen replied:
My hope was that by putting them in order we would create a smooth tree, but this does feel more safe.

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+        # i.e. cells = (np.linspace(-12,12,100),np.linspace(-4,4,50))
+        if isinstance(cells,tuple):
+
+            if isinstance(cells[0],tuple):
+                raise ValueError("LintSampler: Cells must be a single tuple or a list of tuples (i.e. not a tuple of tuples).")
```
I guess at this stage we're insisting that all items of cells are arrays, so could more explicitly type check:
```suggestion
            if not isinstance(cells[0], np.ndarray):
                raise ValueError("LintSampler: Cells must be a tuple of a list of arrays.")
```
But also should this be a ValueError or a TypeError?

michael-petersen replied:
Right, this should be a `TypeError`.

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+        # i.e. cells = [(np.linspace(-12,0,100),np.linspace(-4,0,50)),(np.linspace(0,12,100),np.linspace(0,4,50))]
+        # or a list of grids that have been preconstructed and stacked,
+        # i.e. [np.stack(np.meshgrid(np.linspace(-12,0,100),np.linspace(-4,0,50), indexing='ij'), axis=-1),np.stack(np.meshgrid(np.linspace(0,12,100),np.linspace(0,4,50), indexing='ij'), axis=-1)]
+        if isinstance(cells,list):
```
Make `elif` to avoid catching 1D list (case 0)
```suggestion
        elif isinstance(cells,list):
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+        # i.e. cells = (np.linspace(-12,12,100),np.linspace(-4,4,50))
+        if isinstance(cells,tuple):
+
+            if isinstance(cells[0],tuple):
+                raise ValueError("LintSampler: Cells must be a single tuple or a list of tuples (i.e. not a tuple of tuples).")
```
Alternatively, we could allow the case of tuple of all 1D iterables, i.e., tuple of 1D arrays, tuple of tuples, tuple of lists.
```suggestion
            if not (hasattr(cells[0], "__len__") and not hasattr(cells[0][0], "__len__")):
                raise ValueError("LintSampler: Cells must be a single tuple of 1D sequences (arrays, lists, or tuples).")
```

michael-petersen replied:
...and I think this handling is slightly better for input flexibility.

aneeshnaik reviewed (commented):
Few comments on the different usages of the `cells` argument

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,526 @@
+        ----------
+        self : LintSampler
+            The LintSampler instance.
+        cells : tuple of arrays or list of tuples of arrays
+            If a tuple of arrays, the boundary values for a grid with dimensionality of the length
+            of the tuple. If a list of tuples of arrays, the boundary values for an arbitrary
+            number of grids with dimensionality the length of the tuples.
```
```suggestion
        cells : single array, tuple of arrays or list of tuples of arrays
            If a single array, the boundary values for a 1D grid. If a tuple of arrays, the
            boundary values for a grid with dimensionality of the length of the tuple. If
            a list of tuples of arrays, the boundary values for an arbitrary number of
            grids with dimensionality the length of the tuples.
```

michael-petersen committed:
Apply suggestions from code review

Co-authored-by: Aneesh Naik <48957123+aneeshnaik@users.noreply.github.com>

michael-petersen committed:
ValueError -> TypeError

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,528 @@
+
+        # 0. are we in the 1d case? the input is just a single tuple or 1d array.
+        # i.e. cells = np.linspace(-12,12,100)
+        if hasattr(cells, "__len__") and not hasattr(cells[0], "__len__")
```
Apologies, my previous suggestion was missing a colon:
```suggestion
        if hasattr(cells, "__len__") and not hasattr(cells[0], "__len__"):
```

michael-petersen committed:
Small typo in src/lintsampler/lintsampler.py

Co-authored-by: Aneesh Naik <48957123+aneeshnaik@users.noreply.github.com>

michael-petersen committed:
convert pdf input to take k-dimensional first argument

michael-petersen committed:
add ability to take a single grid input to cells

michael-petersen commented on src/lintsampler/lintsampler.py:
Bit of manipulation required to handle arbitrary-sized grids; I think this works, but may need some further checking.

michael-petersen committed:
remove uneeded case switch

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,547 @@
+
+            # create the grid
+            for d in range(0,self.dim):
+                self.edgearrays.append(cells[d])
```
In case user passes tuple of 1D lists or tuple of 1D tuples (instead of tuple of arrays)
```suggestion
                self.edgearrays.append(np.array(cells[d]))
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,547 @@
+                evalf = self._evaluate_gridded_pdf(funcargs)
+
+                # call the gridded sampler
+                X = _gridsample(self.edgearrays,f=evalf,N_samples=N_samples,seed=self.rng)
```
Below (in `setgrid`), I make the suggestion of always having `edgearrays` as a list, even in the 1D case. Then, we don't need the two cases here.
```suggestion             
            # evaluate the pdf
            evalf = self._evaluate_gridded_pdf(funcargs)

            # call the gridded sampler
            X = _gridsample(*self.edgearrays,f=evalf,N_samples=N_samples,seed=self.rng)
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,547 @@
+
+            # set the arrays
+            # we are being passed the array directly
+            self.edgearrays = np.array(cells)
```
The code in `sample` (lines 227 to 241) can be simplified a little if `edgearrays` is always a list, even in the 1D case.
```suggestion
            self.edgearrays = [np.array(cells)]
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,547 @@
+            The inferred dimensionality of the problem
+        edgearrays : list of numpy arrays
+            The arrays defining the edge of the grids
+        edgedims   : list of integers, optional, used if eval_type == 'gridsample'
+            The shape of the gridsample grid
```
Depending on the case, `edgedims` is sometimes an integer, sometimes a tuple, sometimes a numpy array. I make a code suggestion below to make it always a tuple. Here is an updated description to reflect this:
```suggestion
        edgedims   : tuple of integers, optional, used if eval_type == 'gridsample'
            The shape of the gridsample grid
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,547 @@
+            # set the arrays
+            # we are being passed the array directly
+            self.edgearrays = np.array(cells)
+            self.edgedims = len(cells)
```
```suggestion
            self.edgedims = (len(cells),)
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,547 @@
+            # create the grid
+            for d in range(0,self.dim):
+                self.edgearrays.append(cells[d])
+                self.edgedims[d] = len(cells[d])
```
```suggestion
            self.edgedims = ()

            # create the grid
            for d in range(0,self.dim):
                self.edgearrays.append(cells[d])
                self.edgedims += (len(cells[d]),)
```

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,547 @@
+            if self.dim > 1:
+                evalf = self.pdf(edgegrid,*funcargs).reshape(*self.edgedims)
+            else:
+                evalf = self.pdf(self.edgearrays,*funcargs)
```
Two things here: one is that the 1D case is not yet covered in the non-vectorized route (because edgegrid is not defined in line 148), and the other is that if you accept my suggestion below to make `self.edgearrays` always a list then the 1D case needs to look a bit different. This I think is relatively succinct:
```suggestion
        # create the flattened grid for evaluation in k>1 case
        if self.dim > 1: 
            edgegrid = np.stack(np.meshgrid(*self.edgearrays, indexing='ij'), axis=-1).reshape(np.prod(self.edgedims), self.dim)
        else:
            edgegrid = self.edgearrays[0]

        # reshape the grid: assumes function takes same number of arguments as dimensions
        if self.vectorizedpdf:
            evalf = self.pdf(edgegrid,*funcargs).reshape(*self.edgedims)
```

aneeshnaik reviewed (commented):
A series of suggestions, to achieve the following:
- catch the case where the user passes a tuple of 1D tuples or tuple of 1D lists (rather than tuple of 1D arrays)
- cover the 1D, non-vectorized, gridsample route
- change the assignment of `edgedims` attribute so it is always a tuple (currently sometimes tuple, sometimes array, sometimes integer)
- change assignment of `edgearrays` attribute so it is always a list of arrays (currently it can also be a single array in the 1D case). This change in turn simplifies some code in `sample`

michael-petersen committed:
Enforce list for 1d input

1d inputs are now treated as a list, paralleling the kd case, and simplifying code under the hood.

Co-authored-by: Aneesh Naik <48957123+aneeshnaik@users.noreply.github.com>

michael-petersen committed:
Ensure 1d case is an array

Co-authored-by: Aneesh Naik <48957123+aneeshnaik@users.noreply.github.com>

michael-petersen committed:
add 1d freesample case

aneeshnaik commented on src/lintsampler/lintsampler.py:
```diff
@@ -0,0 +1,535 @@
+            # check if we are being passed grids already
+            if isinstance(cells[0],np.ndarray):
+                _gridsconstructed = True
+
+                # check special case dimensionality: we are being passed multiple 1d grids
+                if (cells[0].shape[0] == cells[0].shape[-1]):
+                    self.dim = 1
+
+            # infer the dimensionality: this will work whether the input is a single cell or list of cells
+            # only proceed with determination if not 1d case
+            if (self.dim == 0):
+                if _gridsconstructed:
+                    self.dim = cells[0].shape[-1]
+                else:
+                    self.dim = len(cells[0])
+
+            # create holding arrays
+            grids                 = [] # the grids constructed from the inputs (no need to expose)
+            self.gridshape        = [] # the shape of the input grids
+            self.ngridentries     = [] # the number of entries in the grid
+            self.nedgegridentries = [] # the number of entries in the edge grid
+            self.edgearrays       = [] # the flattened grids, each [ngridentries,dim]
+
+            # loop through the boundaries, construct the grids, and flatten
+            for ngrid in range(0,self.ngrids):
+               
+                if _gridsconstructed:
+                    # check that the grid has the same dimensionality as the others
+                    if ((self.dim>1) & (cells[ngrid].shape[-1]!=self.dim)):
+                        raise ValueError("LintSampler._setgrid: All input grids must have the same dimensionality.")
+                    
+                    grid = cells[ngrid]
+
+                else:
+                    # check that the dimensions are properly specified for the grid
+                    if len(cells[ngrid])!=self.dim:
+                        raise ValueError("LintSampler._setgrid: All input cells must have the same dimensionality.")
+
+                    # construct the grid
+                    grid = np.stack(np.meshgrid(*cells[ngrid], indexing='ij'), axis=-1)
```
This fixes a couple of issues:
- skips the `meshgrid` business in the 1D case
- previously some types of `cells` input were going down the wrong track

```suggestion
            # infer dimensionality
            # cells = list of tuples of 1D iterables
            if isinstance(cells[0], tuple) and hasattr(cells[0][0], "__len__"):
                self.dim = len(cells[0])
            # cells = list of (k+1)-d grids (i.e. preconstructed k-d grids)
            elif isinstance(cells[0], np.ndarray) and cells[0].ndim > 1:
                _gridsconstructed = True
                self.dim = cells[0].shape[-1]
            # cells = list of 1d iterables
            elif hasattr(cells[0], "__len__") and not hasattr(cells[0][0], "__len__"):
                self.dim = 1
            else:
                raise ValueError("Input cells do not make sense.")

            # create holding arrays
            grids                 = [] # the grids constructed from the inputs (no need to expose)
            self.gridshape        = [] # the shape of the input grids
            self.ngridentries     = [] # the number of entries in the grid
            self.nedgegridentries = [] # the number of entries in the edge grid
            self.edgearrays       = [] # the flattened grids, each [ngridentries,dim]

            # loop through the boundaries, construct the grids, and flatten
            for ngrid in range(0,self.ngrids):
               
                if _gridsconstructed:
                    # check that the grid has the same dimensionality as the others
                    if ((self.dim>1) & (cells[ngrid].shape[-1]!=self.dim)):
                        raise ValueError("LintSampler._setgrid: All input grids must have the same dimensionality.")
                    
                    grid = cells[ngrid]

                else:
                    # check that the dimensions are properly specified for the grid
                    if self.dim > 1 and len(cells[ngrid])!=self.dim:
                        raise ValueError(f"LintSampler._setgrid: All input cells must have the same dimensionality.")

                    # construct the grid
                    if self.dim > 1:
                        grid = np.stack(np.meshgrid(*cells[ngrid], indexing='ij'), axis=-1)
                    else:
                        grid = np.array(cells[ngrid])
```

michael-petersen replied:
I usually like to give the package and function name when returning an error, simply because it tends to make debugging easier. But I like this handling of the inputs, much more smooth with the attribute checking!

michael-petersen committed:
Streamline freesample cell logic

Co-authored-by: Aneesh Naik <48957123+aneeshnaik@users.noreply.github.com>

michael-petersen committed:
specify package in error messages

aneeshnaik committed:
Write tests for lintsampler; remove freesample and gridsampler tests

aneeshnaik committed:
Merge branch 'main' of github.com:michael-petersen/lintsampler into main
